File: leetcode/Algorithms/206.ReverseLinkedList/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

public class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode tail = null;
        while (head != null) {
            ListNode newHead = head.next;
            head.next = tail;
            tail = head;
            head = newHead;
        }
        return tail;
    }
}

File: leetcode/Algorithms/206.ReverseLinkedList/Solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* helper(struct ListNode* head, struct ListNode* newHead) {
    if (head == NULL)
        return newHead;
    struct ListNode* next = head->next;
    head->next = newHead;
    return helper(next, head);
}

struct ListNode* reverseList(struct ListNode* head) {
    return helper(head, NULL);
}
